### PR TITLE
cuDNN Convolution functions Autotuner

### PR DESCRIPTION
**Pull request by Crissman**
PR for @okuta's code to implement cuDNN auto-tune. This PR must be applied in combination with [CuPy PR#624](https://github.com/cupy/cupy/pull/624).

Autotune ability added to Convolution2D, Deconvolution2D, ConvolutionND, and DeconvolutionND.

Documentation for those functions also updated.

Add `chainer.global_config.autotune = True` to enable.

Should new nosetests be added for this?

### CODE REVIEW

**Comment by okuta** [comment]
Code is good.
How about fixing [`configuration.rst`](https://github.com/chainer/chainer/blob/a3db015d960b75d7b621b1232b32405ae50f2f4d/docs/source/reference/core/configuration.rst)?

**Comment by Crissman** [comment]
configuration.rst updated.

**Comment by niboshi** [comment]
#624 (and #664 for stable) have been merged.
When backporting PR, please fix `libcudnn` call because the interface is slightly different from master branch.

**Comment by okuta** [comment]
jenkins, test this please.

**Comment by okuta** [comment]
Please fix tests for `autotune`.

**Comment by Crissman** [comment]
Tests added for autotune forwards and backwards for `convolution_2d`, `convolution_nd`, `deconvolution_2d`, and `deconvolution_nd`.

File: chainer/functions/connection/convolution_2d.py [hunk 1]
```diff
@@ -276,6 +310,10 @@ def forward_gpu(self, inputs):
 
         if configuration.config.cudnn_deterministic:
             algo = libcudnn.CUDNN_CONVOLUTION_BWD_FILTER_ALGO_1
+        elif chainer.global_config.autotune:
```
Comment (okuta) [inline chainer/functions/connection/convolution_2d.py]: Please use `configuration.config.autotune`.

File: chainer/functions/connection/convolution_nd.py [hunk 2]
```diff
@@ -110,19 +111,24 @@ def _forward_cudnn(self, x, W, b):
...
             self.bias_desc = cudnn.create_tensor_descriptor(b[b_index])
 
         # Find cuDNN algorithm to be used.
         workspace_size = cuda.get_max_workspace_size()
         workspace = cuda.cupy.empty((workspace_size,), dtype='b')
-        algo = libcudnn.getConvolutionForwardAlgorithm(
-            handle, x_desc.value, self.filter_desc.value,
-            self.conv_desc.value, y_desc.value, _fwd_pref,
-            workspace_size)
+        if chainer.global_config.autotune:
```
Comment (okuta) [inline chainer/functions/connection/convolution_nd.py]: Please use `configuration.config.autotune`.

File: chainer/functions/connection/convolution_nd.py [hunk 3]
```diff
@@ -230,10 +236,15 @@ def _backward_cudnn(self, x, W, b, gy):
         workspace = cuda.cupy.empty((workspace_size,), dtype='b')
 
         # Compute filter weight gradient.
-        algo = libcudnn.getConvolutionBackwardFilterAlgorithm(
-            handle, x_desc.value, gy_desc.value,
-            self.conv_desc.value, self.filter_desc.value,
-            _bwd_filter_pref, workspace_size)
+        if chainer.global_config.autotune:
```
Comment (okuta) [inline chainer/functions/connection/convolution_nd.py]: Please use `configuration.config.autotune`.

File: chainer/functions/connection/deconvolution_nd.py [hunk 4]
```diff
@@ -132,10 +133,17 @@ def _forward_cudnn(self, x, W, b):
         zero = numpy.array(0, dtype=oz_dtype).ctypes
         workspace_size = cuda.get_max_workspace_size()
         workspace = cuda.cupy.empty((workspace_size,), dtype='b')
-        algo = libcudnn.getConvolutionBackwardDataAlgorithm(
-            handle, self.filter_desc.value, x_desc.value,
-            self.conv_desc.value, y_desc.value, _bwd_data_pref,
-            workspace_size)
+        if chainer.global_config.autotune:
```
Comment (okuta) [inline chainer/functions/connection/deconvolution_nd.py]: Please use `configuration.config.autotune`.

File: chainer/functions/connection/convolution_2d.py [hunk 5]
```diff
@@ -134,19 +164,23 @@ def forward_gpu(self, inputs):
...
             if b is not None:
                 bias_desc = cudnn.create_tensor_descriptor(
                     b[None, :, None, None])
-
             workspace_size = cuda.get_max_workspace_size()
             workspace = cuda.cupy.empty((workspace_size,), dtype='b')
-            algo = libcudnn.getConvolutionForwardAlgorithm(
-                handle, x_desc.value, filter_desc.value,
-                conv_desc.value, y_desc.value, _fwd_pref, workspace_size)
+            if chainer.global_config.autotune:
```
Comment (okuta) [inline chainer/functions/connection/convolution_2d.py]: Please use `configuration.config.autotune`.

File: chainer/functions/connection/deconvolution_2d.py [hunk 6]
```diff
@@ -190,13 +205,16 @@ def forward_gpu(self, inputs):
 
             workspace_size = cuda.get_max_workspace_size()
             workspace = cuda.cupy.empty((workspace_size,), dtype='b')
+
             if configuration.config.cudnn_deterministic:
                 algo = libcudnn.CUDNN_CONVOLUTION_BWD_DATA_ALGO_1
+            elif chainer.global_config.autotune:
```
Comment (okuta) [inline chainer/functions/connection/deconvolution_2d.py]: Please use `configuration.config.autotune`.

File: chainer/functions/connection/convolution_2d.py [hunk 7]
```diff
@@ -369,6 +407,11 @@ def convolution_2d(x, W, b=None, stride=1, pad=0, cover_all=False, **kwargs):
     If ``chainer.configuration.config.cudnn_deterministic`` is ``True`` and
     cuDNN version is >= v3, it forces cuDNN to use a deterministic algorithm.
 
+    Convolution links can use a feature of cuDNN called autotuning, which
+    selects the most efficient CNN algorithm for images of fixed-size,
+    can provide a significant performance boost for fixed neural nets.
+    To enable, set `chainer.global_config.autotune = True`
```
Comment (okuta) [inline chainer/functions/connection/convolution_2d.py]: Please use `chainer.using_config('autotune', value)`.

See also deterministic mode document.
> Use chainer.using_config('cudnn_deterministic', value) context where value is either `True` or `False`.

File: chainer/functions/connection/convolution_nd.py [hunk 8]
```diff
@@ -363,6 +374,11 @@ def convolution_nd(x, W, b=None, stride=1, pad=0, cover_all=False):
         - The ``dtype`` is FP16, FP32 or FP64. (FP16 is only available when
           cuDNN version :math:`\\geq` v3.)
 
+    Convolution links can use a feature of cuDNN called autotuning, which
+    selects the most efficient CNN algorithm for images of fixed-size,
+    can provide a significant performance boost for fixed neural nets.
+    To enable, set `chainer.global_config.autotune = True`
```
Comment (okuta) [inline chainer/functions/connection/convolution_nd.py]: Please use `chainer.using_config('autotune', value)`.

See also deterministic mode document.
> Use chainer.using_config('cudnn_deterministic', value) context where value is either `True` or `False`.

File: chainer/functions/connection/deconvolution_2d.py [hunk 9]
```diff
@@ -302,6 +320,11 @@ def deconvolution_2d(x, W, b=None, stride=1, pad=0, outsize=None, **kwargs):
     If ``chainer.configuration.config.deterministic`` is ``True`` and
     cuDNN version is >= v3, it forces cuDNN to use a deterministic algorithm.
 
+    Convolution links can use a feature of cuDNN called autotuning, which
+    selects the most efficient CNN algorithm for images of fixed-size,
+    can provide a significant performance boost for fixed neural nets.
+    To enable, set `chainer.global_config.autotune = True`
```
Comment (okuta) [inline chainer/functions/connection/deconvolution_2d.py]: Please use `chainer.using_config('autotune', value)`.

See also deterministic mode document.
> Use chainer.using_config('cudnn_deterministic', value) context where value is either `True` or `False`.

File: chainer/links/connection/convolution_2d.py [hunk 10]
```diff
@@ -18,6 +18,11 @@ class Convolution2D(link.Link):
     If ``chainer.configuration.config.deterministic`` is ``True`` and
     cuDNN version is >= v3, it forces cuDNN to use a deterministic algorithm.
 
+    Convolution links can use a feature of cuDNN called autotuning, which
+    selects the most efficient CNN algorithm for images of fixed-size, 
+    can provide a significant performance boost for fixed neural nets.
+    To enable, set `chainer.global_config.autotune = True`
```
Comment (okuta) [inline chainer/links/connection/convolution_2d.py]: Please use `chainer.using_config('autotune', value)`.

See also deterministic mode document.
> Use chainer.using_config('cudnn_deterministic', value) context where value is either `True` or `False`.

File: chainer/functions/connection/convolution_2d.py [hunk 11]
```diff
@@ -26,6 +26,36 @@ def _pair(x):
...
+
+def get_algorithm_bwd_filter(
+        x, dy, dW, conv_param, handle, x_desc, dy_desc, conv_desc, filter_desc,
+        workspace):
+    key = (x.shape, dW.shape, dy.shape, conv_param)
+    if key in _algorithm_bwd_filter:
+        return _algorithm_bwd_filter[key]
+    ret = libcudnn.findConvolutionBackwardFilterAlgorithmEx(
+        handle, x_desc.value, x.data.ptr, dy_desc.value, dy.data.ptr,
+        conv_desc.value, filter_desc.value, dW.data.ptr, 10,
```
Comment (okuta) [inline chainer/functions/connection/convolution_2d.py]: Please fix `10` to `1`. 
I used `10` for only debug.

File: chainer/functions/connection/deconvolution_2d.py [hunk 12]
```diff
@@ -19,6 +19,21 @@
...
+
+
+def get_algorithm(W, dy, dx, conv_param, handle, filter_desc, dy_desc,
+                  conv_desc, dx_desc, workspace):
+    key = (dx.shape, W.shape, dy.shape, conv_param)
+    if key in _algorithm:
+        return _algorithm[key]
+    ret = libcudnn.findConvolutionBackwardDataAlgorithmEx(
+        handle, filter_desc.value, W.data.ptr, dy_desc.value, dy.data.ptr,
+        conv_desc.value, dx_desc.value, dx.data.ptr, 10, workspace.data.ptr,
```
Comment (okuta) [inline chainer/functions/connection/deconvolution_2d.py]: Please fix `10` to `1`. 
I used `10` for only debug.

**Comment by okuta** [comment]
jekins, test this please

**Comment by okuta** [comment]
jenkins, test this please

**Comment by okuta** [comment]
jenkins, test this please.

**Comment by unnonouno** [comment]
jenkins test this please

**Comment by okuta** [comment]
LGTM!

**Pull request merged**